privacy/bolton/model.py

```diff
@@ -0,0 +1,551 @@
+                         "Required method: {0} not found".format(key))
+    if not self._layers_instantiated:  # compile may be called multiple times
+      kernel_intiializer = kwargs.get('kernel_initializer',
+                                      tf.initializers.GlorotUniform)
+      self.output_layer = tf.keras.layers.Dense(
+          self.n_classes,
+          kernel_regularizer=loss.kernel_regularizer(),
+          kernel_initializer=kernel_intiializer(),
+      )
+      # if we don't do regularization here, we require the user to
+      # re-instantiate the model each time they want to change the penalty
+      # weighting
+      self._layers_instantiated = True
+    self.output_layer.kernel_regularizer.l2 = loss.reg_lambda
+    if not isinstance(optimizer, Bolton):
+      optimizer = optimizers.get(optimizer)
+      optimizer = Bolton(optimizer, loss)
+
+    super(BoltonModel, self).compile(optimizer,
+                                     loss=loss,
+                                     metrics=metrics,
+                                     loss_weights=loss_weights,
+                                     sample_weight_mode=sample_weight_mode,
+                                     weighted_metrics=weighted_metrics,
+                                     target_tensors=target_tensors,
+                                     distribute=distribute,
+                                     **kwargs
+                                     )
+
+  # def _post_fit(self, x, n_samples):
+  #   """Implements 1-time weight changes needed for Bolton method.
+  #   In this case, specifically implements the noise addition
+  #   assuming a strongly convex function.
+  #
+  #   Args:
+  #       x: inputs
+  #       n_samples: number of samples in the inputs. In case the number
+  #       cannot be readily determined by inspecting x.
+  #
+  #   Returns:
+  #
+  #   """
+  #   data_size = None
+  #   if n_samples is not None:
+  #     data_size = n_samples
+  #   elif hasattr(x, 'shape'):
+  #     data_size = x.shape[0]
+  #   elif hasattr(x, "__len__"):
+  #     data_size = len(x)
+  #   elif data_size is None:
+  #     if n_samples is None:
+  #       raise ValueError("Unable to detect the number of training "
+  #                        "samples and n_smaples was None. "
+  #                        "either pass a dataset with a .shape or "
+  #                        "__len__ attribute or explicitly pass the "
+  #                        "number of samples as n_smaples.")
+  #   for layer in self.layers:
+  #     # layer.kernel = layer.kernel + self._get_noise(
+  #     #     data_size
+  #     # )
+  #     input_dim = layer.kernel.numpy().shape[0]
+  #     layer.kernel = layer.kernel + self.optimizer.get_noise(
+  #         self.loss,
+  #         data_size,
+  #         input_dim,
+  #         self.n_classes,
+  #         self.class_weight
+  #     )
+
+  def fit(self,
+          x=None,
+          y=None,
+          batch_size=None,
+          epochs=1,
```

Comment:
Unless you are touching a parameter, I suggest you to use **kwargs to forward them.